Twitter_API/code.py

```diff
@@ -0,0 +1,133 @@
+            change_me += "*"*45
+            raise OSError(change_me)
+        pyportal.neo_status((100, 0, 0)) # red = not connected
+        try:
+            esp.connect(secrets)
+        except RuntimeError as error:
+            logger.error("Could not connect to internet: %s", error)
+            logger.error("Retrying in 3 seconds...")
+            time.sleep(3)
+
+#pylint:disable=redefined-outer-name
+def get_bearer_token():
+    """Get the bearer authentication token from twitter."""
+    raw_key = secrets['twitter_api_key'] + ':' + secrets['twitter_secret_key']
+    encoded_key = binascii.b2a_base64(bytes(raw_key, 'utf8'))
+    string_key = bytes.decode(encoded_key)
+    headers = {'Authorization': 'Basic ' + string_key,
+               'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8'}
+    response = requests.post('https://api.twitter.com/oauth2/token',
+                             headers=headers,
+                             data='grant_type=client_credentials')
+    response_dict = json.loads(response.content)
+    if response_dict['token_type'] != 'bearer':
+        halt_and_catch_fire('Wrong token type from twitter: %s', response_dict['token_type'])
+    return response_dict['access_token']
+#pylint:enable=redefined-outer-name
+
+#setup esp interface
+esp32_ready = DigitalInOut(board.ESP_BUSY)
+esp32_gpio0 = DigitalInOut(board.ESP_GPIO0)
+esp32_reset = DigitalInOut(board.ESP_RESET)
+esp32_cs = DigitalInOut(board.ESP_CS)
+spi = busio.SPI(board.SCK, board.MOSI, board.MISO)
+esp = adafruit_esp32spi.ESP_SPIcontrol(spi, esp32_cs, esp32_ready,
+                                       esp32_reset, esp32_gpio0)
+
+# determine the current working directory
+# needed so we know where to find files
+cwd = ("/"+__file__).rsplit('/', 1)[0]
+
+
+# Initialize the pyportal object and let us know what data to fetch and where
+# to display it
+pyportal = PyPortal(status_neopixel=board.NEOPIXEL,
+                    default_bg=cwd + '/twitter_background.bmp',
+                    external_spi=spi,
+                    esp=esp,
+                    text_font=cwd+'/fonts/Helvetica-Bold-16.bdf',
+                    text_position=(20, 60),
+                    text_color=0xFFFFFF,
+                    caption_text='@' + username,
+                    caption_font=cwd+'/fonts/Helvetica-Bold-16.bdf',
+                    caption_position=(5, 210),
+                    caption_color=0x808080)
+
+logger.debug("Connecting to AP %s...", secrets['ssid'])
+connect_esp()
+logger.debug('Connected')
+
+logger.debug('Getting bearer token...')
+bearer_token = get_bearer_token()
+logger.debug('Got it')
+
+headers = {'Authorization': 'Bearer ' + bearer_token}
+
+url = 'https://api.twitter.com/1.1/statuses/user_timeline.json?count=1&screen_name=' + username
+while True:
+    logger.debug('fetching latest tweet from @%s', username)
+    response = requests.get(url,
+                            headers=headers)
```

Comment:
if you need to pass in headers, i think there's a pyportal PR that adds support? lets try that?

Reply:
I tried using just the PyPortal class, but had problems fetching the bearer token before the PyPortal ... ... I think I might have a way around that...

Reply:
ok im interested if there's a way around it. if not, lets 'skip' the pyportal object and just DIY the whole thing with wifimanager & displayio

Reply:
OK... I saw no PRs for a set_headers so I just use `pyportal._headers = '...'`

Reply:
oki!

Reply:
Not exactly pristine software design.. but hey... it works.  Maybe an optional headers param to fetch at some point?

Reply:
if you want to add that to pyportal now i would merge it and then you could use that, seems handy!

Reply:
PR submitted to PyPortal adding get_headers